### PR TITLE
[Op][String][Ref] Extend Concat op reference to support string type

### PR DESCRIPTION
**Pull request by mitruska**
### Details:
 - [Op][String][Ref] Extend Concat op evaluate and reference to support string type
 -  Enablement of CPU fallback to the Concat reference implementation

### Tickets:
 - 131838

### CODE REVIEW

File: src/core/reference/include/openvino/reference/concat.hpp [hunk 1]
```diff
@@ -16,5 +16,12 @@ void concat(const std::vector<const char*>& args,
             const Shape& out_shape,
             int64_t concatenation_axis,
             size_t elem_size);
-}
```
Comment (rkazants) [inline src/core/reference/include/openvino/reference/concat.hpp:19]: let us enable CPU fallback

Comment (mitruska) [reply src/core/reference/include/openvino/reference/concat.hpp:19]: Sure, done

**Review by rkazants — APPROVED** [review]

File: src/core/reference/src/op/concat.cpp [hunk 2]
```diff
@@ -44,5 +44,33 @@ void concat(const std::vector<const char*>& args,
...
+void concat(const std::vector<const std::string*>& args,
+            std::string* out,
+            const std::vector<Shape>& in_shapes,
+            const Shape& out_shape,
+            int64_t concatenation_axis,
+            size_t) {
+    size_t steps = 1;
+    for (int i = 0; i < concatenation_axis; ++i) {
+        steps *= out_shape[i];
+    }
```
Comment (praasz) [inline src/core/reference/src/op/concat.cpp:57]: steps is just shape size
```suggestion
    const size_t steps = concatenation_axis < 0 ? 1 : shape_size(out_shape.begin(), out_shape.begin() + concatenation_axis);
```
`concatenation_axis < 0` could be not required as normalized value is passed

File: src/core/reference/src/op/concat.cpp [hunk 3]
```diff
@@ -44,5 +44,33 @@ void concat(const std::vector<const char*>& args,
...
+    const auto& shape_sizes = calculate_shape_sizes(in_shapes);
+
+    size_t out_offset = 0;
+    for (size_t step = 0; step < steps; ++step) {
```
Comment (praasz) [inline src/core/reference/src/op/concat.cpp:61]: It looks like the main algo for this reference is duplicated. Is possible to make common function and inject only copy function depends on type which base on offset, size etc.?

File: src/core/src/op/concat.cpp [hunk 4]
```diff
@@ -52,34 +52,43 @@ std::shared_ptr<Node> Concat::clone_with_new_inputs(const OutputVector& new_args
...
+                      ov::util::normalize(node->get_axis(), out_shape.size()),
                       outputs.front().get_element_type().size());
+}
+
+bool Concat::evaluate(TensorVector& outputs, const TensorVector& inputs) const {
+    OV_OP_SCOPE(v0_Concat_evaluate);
+    OPENVINO_ASSERT(outputs.size() == 1);
+
+    if (outputs.front().get_element_type() == ov::element::string) {
+        evaluate_concat<std::string>(this, outputs, inputs);
```
Comment (praasz) [inline src/core/src/op/concat.cpp:88]: Instead make the whole evaluate as template (there are some code part which not depends on type) and introduce smaller templates and call them like
```
  auto arg_buff = get_buffers<std::string>(inputs);
   reference::concat(...) // for string
```
  
  and do similar for char* and not duplicate non-template code.

File: src/plugins/intel_cpu/src/nodes/concat.cpp [hunk 5]
```diff
@@ -40,6 +40,9 @@ bool Concat::isSupportedOperation(const std::shared_ptr<const ov::Node>& op, std
             errorMessage = "Node is not an instance of the Concat operation.";
             return false;
         }
+        if (concatOp->get_output_element_type(0) == ov::element::string) {
+            return false;
+        }
```
Comment (praasz) [inline src/plugins/intel_cpu/src/nodes/concat.cpp:45]:
```suggestion
        } else if (concatOp->get_output_element_type(0) == ov::element::string) {
            errorMessage = "Giver error message for not supported type"; // ??
            return false;
        }
```

Comment (mitruska) [reply src/plugins/intel_cpu/src/nodes/concat.cpp:45]: This part is responsible for fallback to the reference, it doesn't throw any error.

**Comment by mitruska** [comment]
Agreed with @praasz to merge current changes, and apply code refactoring suggestions separately.

**Review by praasz — APPROVED** [review]

File: src/core/reference/src/op/concat.cpp (hunk 3 above)
Comment (mitruska) [reply src/core/reference/src/op/concat.cpp:61]: To be applied and reviewed within:
- https://github.com/openvinotoolkit/openvino/pull/23048

File: src/core/src/op/concat.cpp (hunk 4 above)
Comment (mitruska) [reply src/core/src/op/concat.cpp:88]: To be applied and reviewed within:
- https://github.com/openvinotoolkit/openvino/pull/23048

**Pull request merged**